File: 0086.partition_list/solution.py
from utils import ListNode


class Solution:
    def partition(self, head: ListNode, x: int) -> ListNode:
        if not head:
            return head
        ans = low = ListNode(None)
        target = high = ListNode(None)
        while head:
            if head.val < x:
                low.next = head
                low = low.next
            else:
                high.next = head
                high = high.next
            head = head.next
        high.next = None
        low.next = target.next
        return ans.next

File: 0086.partition_list/solution.go
package leetcode0086

import "github.com/Ackerr/Algorithms/utils"


func partition(head *utils.ListNode, x int) *utils.ListNode {
	if head == nil {
		return head
	}

	low := &utils.ListNode{Val: -1}
	ans := low
	high := &utils.ListNode{Val: -1}
	target := high
	for head != nil {
		if head.Val < x {
			low.Next = head
			low = low.Next
		} else {
			high.Next = head
			high = high.Next
		}
		head = head.Next
	}
	high.Next = nil
	low.Next = target.Next
	return ans.Next
}
